src/lib.rs

```diff
@@ -1,51 +1,100 @@
 #![allow(clippy::needless_return, clippy::unreadable_literal)]
+#![no_std]
 
+#[cfg(feature = "metal")]
+#[macro_use]
+extern crate alloc;
+
+#[cfg(feature = "metal")]
+extern crate core;
+
+#[cfg(feature = "std")]
+#[macro_use]
+extern crate std;
+
+#[cfg(any(feature = "std", feature = "metal"))]
 mod arraymap;
+#[cfg(any(feature = "std", feature = "metal"))]
 mod base;
+#[cfg(any(feature = "std", feature = "metal"))]
 mod constraint_matrix;
+#[cfg(any(feature = "std", feature = "metal"))]
 mod decoder;
+#[cfg(any(feature = "std", feature = "metal"))]
 mod encoder;
+mod features_check;
+#[cfg(any(feature = "std", feature = "metal"))]
 mod gf2;
+#[cfg(any(feature = "std", feature = "metal"))]
 mod graph;
+#[cfg(any(feature = "std", feature = "metal"))]
 mod iterators;
+#[cfg(any(feature = "std", feature = "metal"))]
 mod matrix;
+#[cfg(any(feature = "std", feature = "metal"))]
 mod octet;
+#[cfg(any(feature = "std", feature = "metal"))]
 mod octet_matrix;
+#[cfg(any(feature = "std", feature = "metal"))]
```

Comment:
Same comment about all the conditional compilation in this file. I think it's unnecessary?